#' Plots maf summary.
#'
#' @description Plots maf summary.
#' @param maf an \code{\link{MAF}} object generated by \code{\link{read.maf}}
#' @param file If given pdf file will be generated.
#' @param width plot parameter for output file.
#' @param height plot parameter for output file.
#' @param dashboard If FALSE plots simple summary instead of dashboard style.
#' @param titvRaw TRUE. If false instead of raw counts, plots fraction.
#' @param rmOutlier If TRUE removes outlier from boxplot.
#' @param addStat Can be either mean or median. Default NULL.
#' @param showBarcodes include sample names in the top bar plot.
#' @param color named vector of colors for each Variant_Classification.
#' @param textSize font size if showBarcodes is TRUE. Default 2.
#' @param statFontSize font size if addStat is used. Default 3.
#' @param fs base size for text. Default 10.
#' @param top include top n genes dashboard plot. Default 10.
#' @param titvColor colors for SNV classifications.
#' @examples
#' laml.maf <- system.file("extdata", "tcga_laml.maf.gz", package = "maftools")
#' laml <- read.maf(maf = laml.maf, useAll = FALSE)
#' plotmafSummary(maf = laml, addStat = 'median')
#' @return Prints plot.
#' @import RColorBrewer
#' @importFrom cowplot background_grid plot_grid save_plot theme_cowplot
#' @seealso \code{\link{read.maf}} \code{\link{MAF}}
#' @export

plotmafSummary = function(maf, file = NULL, rmOutlier = TRUE, dashboard = TRUE, titvRaw = TRUE,
                          width = 6, height = 5, addStat = NULL, showBarcodes = FALSE, fs = 10,
                          textSize = 2, color = NULL, statFontSize = 3, titvColor = NULL, top = 10){


  if(dashboard){
    #Plot in dashboard style
    pie = FALSE
    gg.summary= dashboard(maf = maf, color = color, rmOutlier = TRUE, titv.color = titvColor, fontSize = fs, sfs = statFontSize, n = top, donut = pie, rawcount = titvRaw)

  }else{

    addStat.opts = c('mean', 'median')

    if(is.null(color)){
      #hard coded color scheme if user doesnt provide any
      col = c(brewer.pal(12,name = "Paired"),brewer.pal(11,name = "Spectral")[1:3],'black')
      names(col) = names = c('Nonstop_Mutation','Frame_Shift_Del','IGR','Missense_Mutation','Silent','Nonsense_Mutation',
                             'RNA','Splice_Site','Intron','Frame_Shift_Ins','Nonstop_Mutation','In_Frame_Del','ITD','In_Frame_Ins','Translation_Start_Site',"Multi_Hit")
    }else{
      col = color
    }

    vcs = getSampleSummary(maf)
    vcs = vcs[,colnames(vcs)[!colnames(x = vcs) %in% c('total', 'Amp', 'Del', 'CNV_total')], with = FALSE]
    #vcs = vcs[,2:(ncol(vcs)-1), with = F]
    #vcs = vcs[order(rowSums(vcs[,2:ncol(vcs), with = F]), decreasing = T)] #order tsbs based on number of mutations
    vcs = vcs[,c(1,order(colSums(x = vcs[,2:(ncol(vcs)), with =FALSE]), decreasing = TRUE)+1), with =FALSE] #order based on most event

    #melt data frame
    vcs.m = data.table::melt(data = vcs, id = 'Tumor_Sample_Barcode')
    colnames(vcs.m) = c('Tumor_Sample_Barcode', 'Variant_Classification', 'N')

    vcs.m$Tumor_Sample_Barcode = factor(vcs.m$Tumor_Sample_Barcode,levels = vcs$Tumor_Sample_Barcode) #reorder tsb levels
    vcs.m$Variant_Classification = factor(x = vcs.m$Variant_Classification, levels = colnames(vcs)) #reorder Variant classification

    #For now keep it at 90
    textAngle = 90

    #top ggplot

    if(!is.null(addStat)){
      if(length(addStat) > 1){
        stop('addStat can only be either mean or median.')
      }

      if(! addStat %in% addStat.opts){
        stop('addStat can only be either mean or median.')
      }

      if(addStat == 'mean'){
        mean.line = round(max(maf@summary[,Mean], na.rm = TRUE), 2)
        df = data.frame(y = c(mean.line), x = as.integer(0.8*nrow(getSampleSummary(maf))), label = c(paste('Mean: ', mean.line, sep='')))

        vcs.gg = ggplot(data = vcs.m, aes(x = Tumor_Sample_Barcode, y = N, fill = Variant_Classification))+geom_bar(stat = 'identity')+scale_fill_manual(values = col)+cowplot::theme_cowplot()+
          theme(axis.ticks = element_blank(), axis.text.x = element_text(angle = textAngle, vjust = 0.5, hjust=1, size = textSize), legend.position = 'none')+ylab('# of Variants')+xlab('Samples')+
          cowplot::background_grid(major = 'xy', minor = 'none')+geom_hline(yintercept = mean.line, linetype = 2)+
          ggrepel::geom_label_repel(inherit.aes = FALSE, data = df, aes(x = x, y = y, label = label), fill = 'gray', fontface = 'bold', color = 'black', box.padding = unit(1, "lines"),
                                    point.padding = unit(1, "lines"), force = 20, size = statFontSize, nudge_y = 2)
      }else{
        med.line = round(max(maf@summary[,Median], na.rm = TRUE), 2)
        df = data.frame(y = c(med.line), x = as.integer(0.8*nrow(getSampleSummary(maf))), label = c(paste('Median: ', med.line, sep='')))

        vcs.gg = ggplot(data = vcs.m, aes(x = Tumor_Sample_Barcode, y = N, fill = Variant_Classification))+geom_bar(stat = 'identity')+scale_fill_manual(values = col)+cowplot::theme_cowplot()+
          theme(axis.ticks = element_blank(), axis.text.x = element_text(angle = textAngle, vjust = 0.5, hjust=1, size = textSize), legend.position = 'none')+ylab('# of Variants')+xlab('Samples')+
          cowplot::background_grid(major = 'xy', minor = 'none')+geom_hline(yintercept = med.line, linetype = 3)+
          ggrepel::geom_label_repel(inherit.aes = FALSE, data = df, aes(x = x, y = y, label = label), fill = 'gray', fontface = 'bold', color = 'black', box.padding = unit(1, "lines"),
                                    point.padding = unit(1, "lines"), force = 20, size = statFontSize, nudge_y = 2)
      }


    }else{
      vcs.gg = ggplot(data = vcs.m, aes(x = Tumor_Sample_Barcode, y = N, fill = Variant_Classification))+geom_bar(stat = 'identity')+scale_fill_manual(values = col)+cowplot::theme_cowplot()+
        theme(axis.ticks = element_blank(), axis.text.x = element_text(angle = textAngle, vjust = 0.5, hjust=1, size = textSize), legend.position = 'none')+ylab('# of Variants')+xlab('Samples')+
        cowplot::background_grid(major = 'xy', minor = 'none')
    }

    if(!showBarcodes){
      vcs.gg = vcs.gg+theme(axis.text.x = element_blank())
    }

    #bottom ggplot
    if(rmOutlier){
      #Get box heights from boxplot.srtas
      boxH = vcs.m[,boxplot.stats(N)$stat[5], by = .(Variant_Classification)]
      colnames(boxH)[ncol(boxH)] = 'boxStat'
      vcs.gg2 = ggplot(data = vcs.m, aes(x = Variant_Classification, y = N, fill = Variant_Classification))+geom_boxplot(outlier.shape = NA)+scale_fill_manual(values = col)+
        cowplot::theme_cowplot()+theme(axis.ticks = element_blank(), axis.text.x = element_blank(), legend.position = 'bottom')+
        theme(legend.key.size = unit(0.5, "cm"), legend.title = element_blank(), legend.text = element_text(size = 8))+ylim(0, max(boxH$boxStat)+5)+cowplot::background_grid(major = 'xy', minor = 'none')
    } else{
      vcs.gg2 = ggplot(data = vcs.m, aes(x = Variant_Classification, y = N, fill = Variant_Classification))+geom_boxplot()+scale_fill_manual(values = col)+cowplot::theme_cowplot()+
        theme(axis.ticks = element_blank(), axis.text.x = element_blank(), legend.position = 'bottom')+
        theme(legend.key.size = unit(0.5, "cm"), legend.title = element_blank(), legend.text = element_text(size = 8))+cowplot::background_grid(major = 'xy', minor = 'none')
    }

    #organize plots
    gg.summary = cowplot::plot_grid(vcs.gg, vcs.gg2, nrow = 2, ncol = 1)

  }

  #print to file if specified.
  if(!is.null(file)){
    cowplot::save_plot(filename = paste(file, 'pdf', sep = '.'), plot = gg.summary, base_width = width, base_height = height, base_aspect_ratio = 1.5)
  }

  print(gg.summary)

}
