#' Subset MAF
#'
#' @description Subsets MAF based on given conditions.
#' @param maf an MAF object generated by \code{\link{read.maf}}
#' @param includeSyn to include sysnonymous variants in output
#' @param tsb subset by these samples (Tumor Sample Barcodes)
#' @param genes subset by these genes
#' @param fields include only these fields along with necessary fields in the output
#' @param isTCGA Is input MAF file from TCGA source.
#' @param query query string. e.g, "Variant_Classification == 'Missense_Mutation'" returns only Missense variants.
#' @param mafObj returns output as MAF class \code{\link{MAF-class}}. Default FALSE
#' @return subset table or an object of class \code{\link{MAF-class}}
#' @seealso \code{\link{getFields}}
#' @examples
#' laml.maf <- system.file("extdata", "tcga_laml.maf.gz", package = "maftools")
#' laml <- read.maf(maf = laml.maf, removeSilent = TRUE, useAll = FALSE)
#' ##Select all Splice_Site mutations from DNMT3A and NPM1
#' subsetMaf(maf = laml, genes = c('DNMT3A', 'NPM1'),
#' query = "Variant_Classification == 'Splice_Site'")
#' ##Select all variants with VAF above 30%
#' subsetMaf(maf = laml, query = "i_TumorVAF_WU > 30")
#' ##Extract data for samples 'TCGA.AB.3009' and 'TCGA.AB.2933' but only include vaf filed.
#' subsetMaf(maf = laml, tsb = c('TCGA.AB.3009', 'TCGA.AB.2933'), fields = 'i_TumorVAF_WU')
#' @export

subsetMaf = function(maf, includeSyn = FALSE, tsb = NULL, genes = NULL, fields = NULL, query = NULL, mafObj = FALSE, isTCGA = FALSE){

  #Synonymous variants
  maf.silent = maf@maf.silent
  #Main data
  maf.dat = maf@data

  #in case user read maf without removing silent variants, remove theme here.
  silent = c("3'UTR", "5'UTR", "3'Flank", "Targeted_Region", "Silent", "Intron",
             "RNA", "IGR", "Splice_Region", "5'Flank", "lincRNA")
  maf.dat = maf.dat[!Variant_Classification %in% silent] #Remove silent variants from main table

  maf.dat = rbind(maf.dat, maf.silent, fill = TRUE)

  #Select
  if(!is.null(tsb)){

    tsb = gsub(pattern = '-', replacement = '.', x = tsb)

    if(isTCGA){
      tsb = substr(x = tsb, start = 1, stop = 12)
    }
    maf.dat = maf.dat[Tumor_Sample_Barcode %in% tsb,]
  }

  if(!is.null(genes)){
    maf.dat = maf.dat[Hugo_Symbol %in% genes, ]
  }

  if(!is.null(query)){
    maf.dat = maf.dat[eval(parse(text=query))]
  }

  default.fields = c('Hugo_Symbol', 'Chromosome', 'Start_Position', 'End_Position', 'Reference_Allele', 'Tumor_Seq_Allele1','Tumor_Seq_Allele2','Variant_Classification', 'Variant_Type', 'Tumor_Sample_Barcode') #necessary fields.

  if(!is.null(fields)){
    default.fields = c(default.fields, fields)
    default.fields = unique(default.fields)

    maf.dat = maf.dat[,default.fields, with = FALSE]
  }


  maf.silent = maf.dat[Variant_Classification %in% silent]

  if(!includeSyn){
    maf.dat = maf.dat[!Variant_Classification %in% silent]
  }


  if(mafObj){

      mafSummary = summarizeMaf(maf.dat, chatty = FALSE)
      oncomat = createOncoMatrix(maf.dat, chatty = FALSE)

      m = MAF(data = maf.dat, variants.per.sample = mafSummary$variants.per.sample, variant.type.summary = mafSummary$variant.type.summary,
              variant.classification.summary = mafSummary$variant.classification.summary,gene.summary = mafSummary$gene.summary,
              oncoMatrix = oncomat$oncomat, numericMatrix = oncomat$nummat, summary = mafSummary$summary,
              classCode = oncomat$vc, maf.silent = maf.silent)

    return(m)
  }else{
    return(maf.dat)
  }
}
