#' Extract single 5' and 3' bases flanking the mutated site.
#' @details Extracts immediate 5' and 3' bases flanking the mutated site and classifies them into 96 substitution classes.
#' This function loads reference genome into memeory. Typical human geneome occupies a peak memory of ~3 gb while extracting bases.
#'
#' @param maf an \code{\link{MAF}} object generated by \code{\link{read.maf}}
#' @param ref_genome faidx indexed refrence fasta file.
#' @param prefix Prefix to add or remove from contig names in MAF file.
#' @param add If prefix is used, default is to add prefix to contig names in MAF file. If false prefix will be removed from contig names.
#' @param ignoreChr Chromsomes to remove from analysis. e.g. chrM
#' @param useSyn Logical. Whether to include synonymous variants in analysis. Defaults to FALSE.
#' @return A matrix of dimension nx96, where n is the number of samples in the MAF.
#' @examples
#' \dontrun{
#' laml.tnm <- trinucleotideMatrix(maf = laml, ref_genome = 'hg19.fa',
#' prefix = 'chr', add = TRUE, useSyn = TRUE)
#' }
#'
#' @importFrom VariantAnnotation getSeq seqlevels
#' @importFrom Biostrings subseq
#' @importFrom Rsamtools FaFile
#' @seealso \code{\link{extractSignatures}}
#' @export

trinucleotideMatrix = function(maf, ref_genome, prefix = NULL, add = TRUE, ignoreChr = NULL, useSyn = FALSE){

  #suppressPackageStartupMessages(require('VariantAnnotation', quietly = TRUE))
  #suppressPackageStartupMessages(require('Biostrings', quietly = TRUE))

  #Synonymous variants
  maf.silent = maf@maf.silent
  #Main data
  maf = maf@data

  #in case user read maf without removing silent variants, remove theme here.
  silent = c("3'UTR", "5'UTR", "3'Flank", "Targeted_Region", "Silent", "Intron",
             "RNA", "IGR", "Splice_Region", "5'Flank", "lincRNA")
  maf = maf[!Variant_Classification %in% silent] #Remove silent variants from main table

  if(useSyn){
    maf = rbind(maf, maf.silent, fill = TRUE)
  }

  #one bp up and down.
  up = down = 1

  #reate a reference to an indexed fasta file.
  ref = Rsamtools::FaFile(file = ref_genome)

  #Remove unwanted contigs
  if(!is.null(ignoreChr)){
    maf = maf[!maf$Chromosome %in% ignoreChr]
  }

  if(!is.null(prefix)){
    if(add){
      maf$Chromosome = paste(prefix,maf$Chromosome, sep = '')
    }else{
      maf$Chromosome = gsub(pattern = prefix, replacement = '', x = maf$Chromosome, fixed = TRUE)
    }
  }

  #seperate snps and indels
  maf.snp = maf[Variant_Type == 'SNP']
  if(nrow(maf) == 0){
    stop('No more single nucleotide variants left after filtering for SNP in Variant_Type field.')
  }
  #maf.rest = maf[!maf$Variant_Type %in% 'SNP']

  #get unique Chromosome names from maf
  chrs = unique(maf.snp$Chromosome)

  #read fasta
  message('reading fasta (this might take few minutes)..')
  ref = VariantAnnotation::getSeq(x = ref)

  #extract contigs from reference fasta
  seq.lvl = VariantAnnotation::seqlevels(ref)

  chrs.missing = chrs[!chrs %in% seq.lvl]

  #validation
  if(length(chrs.missing) > 0){
    message("contigs in fasta file:")
    print(seq.lvl)
    message("contigs in maf:")
    print(chrs)
    message('missing reference contigs from fasta file.')
    print(chrs.missing)
    stop("Contig names in maf must match to contig names in reference fasta. Use prefix to add or remove prefixes from maf if its necessary. Use ignoreChr to ignore missing contigs.")
  }


  extract.tbl = data.table::data.table(Chromosome = maf.snp$Chromosome, Start = maf.snp$Start_Position-1, End = maf.snp$Start_Position+1,
                           Reference_Allele = maf.snp$Reference_Allele, Tumor_Seq_Allele2 = maf.snp$Tumor_Seq_Allele2, Tumor_Sample_Barcode = maf.snp$Tumor_Sample_Barcode)

  message('Extracting adjacent bases..')
  ss = Biostrings::subseq(x = ref[extract.tbl[,Chromosome]], start = extract.tbl[,Start] , end = extract.tbl[,End])
  extract.tbl[,trinucleotide:= as.character(ss)]
  extract.tbl[,Substitution:=paste(extract.tbl$Reference_Allele, extract.tbl$Tumor_Seq_Allele2, sep='>')]

  conv = c("T>C", "T>C", "C>T", "C>T", "T>A", "T>A", "T>G", "T>G", "C>A", "C>A", "C>G", "C>G")
  names(conv) = c('A>G', 'T>C', 'C>T', 'G>A', 'A>T', 'T>A', 'A>C', 'T>G', 'C>A', 'G>T', 'C>G', 'G>C')
  extract.tbl$SubstitutionType = conv[extract.tbl$Substitution]
  # suppressWarnings( extract.tbl[,SubstitutionType:= as.character(factor(x = extract.tbl$Substitution, levels = c('A>G', 'T>C', 'C>T', 'G>A', 'A>T', 'T>A', 'A>C', 'T>G', 'C>A', 'G>T', 'C>G', 'G>C'),
  #                                                                       labels = c("T>C", "T>C", "C>T", "C>T", "T>A", "T>A", "T>G", "T>G", "C>A", "C>A", "C>G", "C>G")))])

  type = paste(substr(x = as.character(extract.tbl$trinucleotide), 1, 1),'[',extract.tbl$SubstitutionType, ']', substr(as.character(extract.tbl$trinucleotide), 3, 3), sep='')
  extract.tbl[,SomaticMutationType:= type]

  extract.tbl.summary = extract.tbl[,.N , by = list(Tumor_Sample_Barcode, SomaticMutationType)]

  colOrder = c("A[C>A]A", "A[C>A]C", "A[C>A]G", "A[C>A]T", "C[C>A]A", "C[C>A]C",
               "C[C>A]G", "C[C>A]T", "G[C>A]A", "G[C>A]C", "G[C>A]G", "G[C>A]T",
               "T[C>A]A", "T[C>A]C", "T[C>A]G", "T[C>A]T", "A[C>G]A", "A[C>G]C",
               "A[C>G]G", "A[C>G]T", "C[C>G]A", "C[C>G]C", "C[C>G]G", "C[C>G]T",
               "G[C>G]A", "G[C>G]C", "G[C>G]G", "G[C>G]T", "T[C>G]A", "T[C>G]C",
               "T[C>G]G", "T[C>G]T", "A[C>T]A", "A[C>T]C", "A[C>T]G", "A[C>T]T",
               "C[C>T]A", "C[C>T]C", "C[C>T]G", "C[C>T]T", "G[C>T]A", "G[C>T]C",
               "G[C>T]G", "G[C>T]T", "T[C>T]A", "T[C>T]C", "T[C>T]G", "T[C>T]T",
               "A[T>A]A", "A[T>A]C", "A[T>A]G", "A[T>A]T", "C[T>A]A", "C[T>A]C",
               "C[T>A]G", "C[T>A]T", "G[T>A]A", "G[T>A]C", "G[T>A]G", "G[T>A]T",
               "T[T>A]A", "T[T>A]C", "T[T>A]G", "T[T>A]T", "A[T>C]A", "A[T>C]C",
               "A[T>C]G", "A[T>C]T", "C[T>C]A", "C[T>C]C", "C[T>C]G", "C[T>C]T",
               "G[T>C]A", "G[T>C]C", "G[T>C]G", "G[T>C]T", "T[T>C]A", "T[T>C]C",
               "T[T>C]G", "T[T>C]T", "A[T>G]A", "A[T>G]C", "A[T>G]G", "A[T>G]T",
               "C[T>G]A", "C[T>G]C", "C[T>G]G", "C[T>G]T", "G[T>G]A", "G[T>G]C",
               "G[T>G]G", "G[T>G]T", "T[T>G]A", "T[T>G]C", "T[T>G]G", "T[T>G]T"
  )

  #colOrderClasses = rep(c('C>A', 'C>G', 'C>T', 'T>A', 'T>C', 'T>G'), each = 16)

  conv.mat = as.data.frame(data.table::dcast(extract.tbl.summary, formula = Tumor_Sample_Barcode~SomaticMutationType, fill = 0, value.var = 'N'))
  #head(conv.mat)
  rownames(conv.mat) = conv.mat[,1]
  #head(conv.mat)
  conv.mat = conv.mat[,-1]

  #conv.mat = t(t(conv.mat)[colOrder,])
  #Check for missing somatic mutation types (this is possible for cancer types with low mutation rate or for a cohort with lesser samples)
  colOrder.missing = colOrder[!colOrder %in% colnames(conv.mat)]
  #If any missing add them with zero counts
  if(length(colOrder.missing) > 0){
    zeroMat = as.data.frame(matrix(data = 0, nrow = nrow(conv.mat), ncol = length(colOrder.missing)))
    colnames(zeroMat) = colOrder.missing
    conv.mat = cbind(conv.mat, zeroMat)
  }

  conv.mat = as.matrix(conv.mat[,match(colOrder, colnames(conv.mat))]) #organize columns according to colOrder

  #Set NAs to zeros if any
  conv.mat[is.na(conv.mat)] = 0
  message(paste('matrix of dimension ', nrow(conv.mat), 'x', ncol(conv.mat), sep=''))
  return(conv.mat)
}
